### ISSUE TITLE
Can't connect with a 'Matching client config not found' error since 1.3.1

### ISSUE THREAD

**Issue by trustin** [yuezk/GlobalProtect-openconnect#85]
I can't connect to my GlobalProtect VPN anymore since I upgraded from 1.3.0 to 1.3.1. 

```
08:32:30.731 INFO  [361681] [gpclient::helper::parseGatewayResponse@50] Start parsing the gateway response...
08:32:30.731 INFO  [361681] [gpclient::helper::parseGatewayResponse@51] The gateway response is: <?xml version="1.0" encoding="utf-8"?><jnlp><application-desc><argument>(null)</argument><argument>[REDACTED]</argument><argument>[REDACTED]</argument><argument>[REDACTED]</argument><argument>[REDACTED]</argument><argument>[REDACTED]</argument><argument>[REDACTED]</argument><argument>%28empty_domain%29</argument><argument>(null)</argument><argument></argument><argument></argument><argument></argument><argument>tunnel</argument><argument>-1</argument><argument>4100</argument><argument></argument><argument></argument><argument></argument><argument></argument><argument>4</argument><argument>unknown</argument><argument></argument></application-desc></jnlp>
08:32:30.731 INFO  [361681] [GPClient::onGatewaySuccess@373] Gateway login succeeded, got the cookie authcookie=[REDACTED]&portal=[REDACTED]&user=[REDACTED]&domain=%2528empty_domain%2529&preferred-ip=&computer=infinity
08:32:30.734 INFO  [361681] [GPClient::onVPNLogAvailable@489] Start process with arugments: --protocol=gp -u  -C authcookie=[REDACTED]&portal=[REDACTED]&user=[REDACTED]&domain=%2528empty_domain%2529&preferred-ip=&computer=infinity [REDACTED]
08:32:30.734 INFO  [361681] [GPClient::onVPNLogAvailable@489] Openconnect started successfully, PID=362122
08:32:30.739 INFO  [361681] [GPClient::onVPNLogAvailable@489] POST https://[REDACTED]/ssl-vpn/getconfig.esp
08:32:30.901 INFO  [361681] [GPClient::onVPNLogAvailable@489] Connected to [REDACTED]:443
08:32:30.927 INFO  [361681] [GPClient::onVPNLogAvailable@489] SSL negotiation with [REDACTED]
08:32:31.251 INFO  [361681] [GPClient::onVPNLogAvailable@489] Connected to HTTPS on [REDACTED] with ciphersuite (TLS1.2)-(RSA)-(AES-256-GCM)
08:32:31.419 INFO  [361681] [GPClient::onVPNLogAvailable@489] Matching client config not found
Creating SSL connection failed
08:32:31.420 INFO  [361681] [GPClient::onVPNLogAvailable@489] Openconnect process exited with code 1 and exit status NormalExit
```

What could be causing this?

**Comment by yuezk** [comment]
Could please fill a value for the “clientos” field in the settings dialog? Possible value could be `Linux` or `Windows`.

https://github.com/yuezk/GlobalProtect-openconnect#passing-the-custom-parameters-to-openconnect-cli

**Comment by yuezk** [comment]
Hi @trustin,

In 1.3.1, I made the value of `clientos` configurable,  the previous value is `Linux`, which was hardcoded in the code. While some users reported (#42) that they want to make the `clientos` configurable because their VPN server doesn't support the value of `Linux`. So I removed the hardcoded value and made it configurable, if the value is empty, then I won't append this value. And it works well with my VPN server.

So, please try the solution in the previous comment, and let me know if it works as expected. Thanks.

**Comment by trustin** [comment]
Thanks for a quick reply, @yuezk. I explicitly specified `Linux` in the 'Value of "clientof"' field but the problem is still there.

**Comment by yuezk** [comment]
> Should I also fill the 'Custom parameters' field? 
I think there is no need to fill the `Custom parameters`.

Is that possible to build the 1.3.0 by yourself on your machine to see whether it is reproducible in 1.3.0? You can follow the following steps to compile it.

```sh
git clone https://github.com/yuezk/GlobalProtect-openconnect.git
cd GlobalProtect-openconnect
git checkout v1.3.0
git submodule update --init

# qmake or qmake-qt5
qmake CONFIG+=release
make
sudo make install
```

**Comment by trustin** [comment]
I rebuilt 1.3.0 from AUR and the newly built 1.3.0 works just fine. 1.3.1 and 1.3.2 don't.

```
❯ git clone https://aur.archlinux.org/globalprotect-openconnect.git
Cloning into 'globalprotect-openconnect'...
remote: Enumerating objects: 84, done.
remote: Counting objects: 100% (84/84), done.
remote: Compressing objects: 100% (61/61), done.
remote: Total 84 (delta 21), reused 84 (delta 21), pack-reused 0
Unpacking objects: 100% (84/84), 437.11 KiB | 397.00 KiB/s, done.
❯ cd globalprotect-openconnect
❯ git reset --hard HEAD~~
HEAD is now at 88f588d Release v1.3.0
❯ git log -1
...
Release v1.3.0
❯ makepkg -s
==> Making package: globalprotect-openconnect 1.3.0-1 (Fri 03 Sep 2021 11:45:31 PM KST)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
...
==> Validating source files with sha256sums...
    GlobalProtect-openconnect-1.3.0.tar.gz ... Passed
    v3.0.19.tar.gz ... Passed
    1.1.5.tar.gz ... Passed
==> Extracting sources...
  -> Extracting GlobalProtect-openconnect-1.3.0.tar.gz with bsdtar
  -> Extracting v3.0.19.tar.gz with bsdtar
  -> Extracting 1.1.5.tar.gz with bsdtar
==> Starting prepare()...
==> Starting build()...
...
```

**Comment by trustin** [comment]
The PKGBUILD difference from the two seems so trivial:

```diff
diff --git PKGBUILD PKGBUILD
index 0714a53..053a94c 100644
--- PKGBUILD
+++ PKGBUILD
@@ -2,7 +2,7 @@
 
 pkgname=globalprotect-openconnect
 _gitname=GlobalProtect-openconnect
-pkgver=1.3.0
+pkgver=1.3.1
 pkgrel=1
 pkgdesc="A GlobalProtect VPN client (GUI) for Linux based on Openconnect and built with Qt5, supports SAML auth mode."
 arch=(x86_64 aarch64)
@@ -17,7 +17,7 @@ source=(
 )
 
 sha256sums=(
-    '0ec932ae988cacf8900af294461fc7e864e645d97ed4bb5662110b113053957b'
+    '0c539dbb781261705c6c688d06951cfbb744033f713d9d8ed4e10e3463d90f4b'
     '9405fd259288b2a862e91e5135bccee936f0438e1b32c13603277132309d15e0'
     '6c80b4701183d2415bec927e1f5ca9b1761b3b5c65d3e09fb29c743e016d5609'
 );

```

**Comment by trustin** [comment]
I even rebuilt 1.3.2 in the same shell session just in case there was any environmental change when I built it before, but problem is still there. It seems like there's something between 1.3.0 and 1.3.1 we don't know yet. :thinking: Please feel free to let me know if you want me to try more things.

**Comment by yuezk** [comment]
Thanks for your feedback? Could you please provide the log files of 1.3.0 and 1.3.1? Thanks.

The log files located at `~/.cache/GlobalProtect-openconnect/gpclient.log`

**Comment by trustin** [comment]
I compared the log files from 1.3.0 and 1.3.1 and found the following differences, although I'm not sure if they make any difference:

- The order of query parameters. In 1.3.0, clientos is right next to os-version, but it's the last query parameter in 1.3.1.
- The XML gateway response in 1.3.0 contains a Base64-encoded string but the response in 1.3.1 doesn't. They both contain the auth cookie though.

**Comment by yuezk** [comment]
Another question:  which authentication mode are you using? The username/password authentication mode with a normal login dialog (shown below) or the SAML authentication mode with a launched WebView?

![Screenshot_20210904_163503](https://user-images.githubusercontent.com/3297602/132088463-46ddbc3d-d14b-42e3-8ad2-af0d53488749.png)

**Comment by yuezk** [comment]
Nevermind. I just got your log files, thank you very much.

**Comment by yuezk** [comment]
The reason should be that the `clientos` parameter is missed in the gateway login request. From the log files, there are 2 occurrences of `clientos` in 1.3.0, while only 1 occurrence in 1.3.1.

Fixed in #87 and released in 1.3.3, please update and try again. Thanks.

**Comment by trustin** [comment]
> Fixed in #87 and released in 1.3.3, please update and try again. Thanks.

I can successfully connect to the VPN with 1.3.1. Thank you so much for fixing this very quickly. :bow: :confetti_ball:
